verifyGC failures related to AudioBufferSourceNode

basuke opened the pull request:
#### e1d0084801c2039a5afe13fd39434897c3920399
<pre>
verifyGC failures related to AudioBufferSourceNode
<a href="https://bugs.webkit.org/show_bug.cgi?id=279596">https://bugs.webkit.org/show_bug.cgi?id=279596</a>
<a href="https://rdar.apple.com/133248373">rdar://133248373</a>

Reviewed by Keith Miller.

Seeing verifyGC failure because of the write barrier missing when touching the AudioBuffer&apos;s property.

* Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
(WebCore::AudioBuffer::getChannelData):
* Source/WebCore/Modules/webaudio/AudioBuffer.h:

Canonical link: <a href="https://commits.webkit.org/283576@main">https://commits.webkit.org/283576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fbc55bc373fe16001d9fd495b6b4db7ccb4adce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66657 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70691 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17790 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68775 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17555 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53407 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12001 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69724 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42380 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57679 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34076 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39051 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15070 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16144 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60966 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15411 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72393 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14776 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60736 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/snapshot-containing-block-includes-scrollbar-gutter.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57743 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61069 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14721 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8735 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2352 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41839 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42916 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44099 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42659 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
```diff
@@ -187,7 +189,7 @@ ExceptionOr<JSC::JSValue> AudioBuffer::getChannelData(JSDOMGlobalObject& globalO
     if (globalObject.worldIsNormal()) {
         if (!m_channelWrappers[channelIndex])
-            m_channelWrappers[channelIndex].setWeakly(constructJSArray());
+            m_channelWrappers[channelIndex].set(globalObject.vm(), wrapper(), constructJSArray());
```
Who creates the `wrapper()`? Is it just `nullptr` since no one called `ScriptWrappable::setWrapper`?

cdumez replied:
setWrapper() should be getting called:
```
Source/WebCore/bindings/js/JSDOMWrapperCache.h:    domObject->setWrapper(wrapper, wrapperOwner, &world);
```

It gets created when JS calls one of our APIs the first time and we need to return said wrapper to them.

kmiller68 replied:
But we didn't need a wrapper before? So what changed to cause the wrapper now? Is there some metaprogramming to tell `AudioBuffer` now is a subclass of `ScriptWrappable` in the bindings?

cdumez replied:
We always had a wrapper (There was always a AudioBuffer.idl and thus an auto-generated JSAudioBuffer wrapper class), now the wrapper is inline though so the C++ implementation can have easy access to it. `ScriptWrappable` is just a way for C++ classes to get easy access to their JS wrapper I believe.

kmiller68 replied:
Ah, that's the part I was missing. Thanks.

cdumez replied:
>Is there some metaprogramming to tell AudioBuffer now is a subclass of ScriptWrappable in the bindings?

I think we used to add `[ScriptWrappable]` to the IDL to make things work. However, it doesn't seem to be a thing anymore. I suspect it now work with some sort of meta-programming but I'll try and find out.

@basuke It would be good to validate with logging to make sure it doesn't always return null :)

cdumez replied:
It's because of these overloads:
```
bool setInlineCachedWrapper(DOMWrapperWorld&, void*, JSDOMObject*, JSC::WeakHandleOwner*);
bool setInlineCachedWrapper(DOMWrapperWorld&, ScriptWrappable*, JSDOMObject* wrapper, JSC::WeakHandleOwner* wrapperOwner);
bool setInlineCachedWrapper(DOMWrapperWorld&, JSC::ArrayBuffer*, JSC::JSArrayBuffer* wrapper, JSC::WeakHandleOwner* wrapperOwner);
```

If you subclass `ScriptWrappable`, then it will call the second overload, which will call `setWrapper()`. Otherwise, it will still create a JS wrapper but not store it on the ScriptWrappable.

basuke replied:
Thanks for clarifying.

kmiller68 reviewed (approved):
r=me.

basuke committed:
verifyGC failures related to AudioBufferSourceNode
https://bugs.webkit.org/show_bug.cgi?id=279596
rdar://133248373

Reviewed by Keith Miller.

Seeing verifyGC failure because of the write barrier missing when touching the AudioBuffer's property.

* Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
(WebCore::AudioBuffer::getChannelData):
* Source/WebCore/Modules/webaudio/AudioBuffer.h:

Canonical link: https://commits.webkit.org/283576@main
